### PR TITLE
rfbserver: add setDesktopSize / ExtendedDesktopSize support

### PR DESCRIPTION
**Pull request by maxnet**
Adds support / hook functions for setDesktopSize command and ExtendedDesktopSize pseudo-encoding.
This is used by clients like noVNC (when using the "remote resizing"
option) to indicate to the server the desired framebuffer size.

### CODE REVIEW

**Comment by bk138** [comment]
Thanks for that, I'll review ASAP!

**Comment by bk138** [comment]
note for me: https://github.com/rfbproto/rfbproto/blob/master/rfbproto.rst#setdesktopsize

File: libvncserver/rfbserver.c [hunk 1]
```diff
@@ -2705,6 +2724,74 @@ rfbProcessClientNormalMessage(rfbClientPtr cl)
...
+            extDesktopScreens[i].x = Swap16IfLE(extDesktopScreens[i].x);
+            extDesktopScreens[i].y = Swap16IfLE(extDesktopScreens[i].y);
+            extDesktopScreens[i].width = Swap16IfLE(extDesktopScreens[i].width);
+            extDesktopScreens[i].height = Swap16IfLE(extDesktopScreens[i].height);
+            extDesktopScreens[i].flags = Swap32IfLE(extDesktopScreens[i].flags);
+        }
+        msg.sdm.width = Swap16IfLE(msg.sdm.width);
+        msg.sdm.height = Swap16IfLE(msg.sdm.height);
+
+        rfbLog("Client requested resolution change to (%dx%d)\n", msg.sdm.width, msg.sdm.height);
```
Comment (bk138) [inline libvncserver/rfbserver.c]: How do we deal with the first paragraph of https://github.com/rfbproto/rfbproto/blob/master/rfbproto.rst#7410setdesktopsize

> Requests a change of desktop size. This message is an extension and may only be sent if the client has previously received an ExtendedDesktopSize rectangle.

Ignore?

File: libvncserver/rfbserver.c [hunk 2]
```diff
@@ -3412,6 +3507,101 @@ rfbSendNewFBSize(rfbClientPtr cl,
     return TRUE;
 }
 
+/*
+ * Send ExtDesktopSize pseudo-rectangle. This message is used:
+ * - to tell the client to change its framebuffer size
+ * - at the start of the session to inform the client we support size changes through setDesktopSize
```
Comment (bk138) [inline libvncserver/rfbserver.c]: I can't find `rfbSendExtDesktopSize` at session start in the code.

Comment (bk138) [reply libvncserver/rfbserver.c]: Or is it: 
* rfbEncodingExtDesktopSize received in SetEncodings from client
*  cl->useExtDesktopSize = TRUE; 
* rfbSendExtDesktopSize() triggered later
?

**Comment by maxnet** [comment]
>I can't find rfbSendExtDesktopSize at session start in the code.

"Start of the session" should not be taken literally.
It is not send straight away the moment the client connects, but at the beginning of session once the client requests its first framebuffer update, provided it has indicated it supports the pseudo encoding.

Sending is triggered by setting newFBSizePending to TRUE when the incremental number is zero:

```
       if (!msg.fur.incremental) {
	    sraRgnOr(cl->modifiedRegion,tmpRegion);
	    sraRgnSubtract(cl->copyRegion,tmpRegion);
            if (cl->useExtDesktopSize)
                cl->newFBSizePending = TRUE;
```


>How do we deal with the first paragraph of https://github.com/rfbproto/rfbproto/blob/master/rfbproto.rst#7410setdesktopsize

It only says the client should not be sending the command before it knows the server has support for it. 
Does not say a server must deny if client does send it earlier.
Are you suggesting we deny the change in that case anyway?

>setDesktopSizeHook is not optional?

It is optional, but if you do not provide it all requests to change resolution will be denied.

File: libvncserver/rfbserver.c (hunk 2 above)
Comment (bk138) [reply libvncserver/rfbserver.c]: thanks for the explanation

File: rfb/rfb.h [hunk 3]
```diff
@@ -695,11 +705,17 @@ typedef struct _rfbClientRec {
 #ifdef LIBVNCSERVER_HAVE_LIBPTHREAD
     int pipe_notify_client_thread[2];
 #endif
+
+    rfbBool useExtDesktopSize;
+    int requestedDesktopSizeChange;
+    int lastDesktopSizeChangeError;
```
[the remark this answers is not in the document]
Comment (bk138) [reply rfb/rfb.h]: just did this for you

**Comment by bk138** [comment]
Closed by 8e41510f4a9d449dd228e5b3e29732882f7f5df6

**Pull request closed without merging**